sympy/matrices/common.py

```diff
@@ -486,6 +486,33 @@ def row_join(self, other):
                 "`self` and `rhs` must have the same number of rows.")
         return self._eval_row_join(other)
 
+    def diagonal(self, k):
+        """Return a tuple of the the kth diagonal.
```

Comment:
Specify what the "kth diagonal" means here. How about:

"The kth diagonal is offset by k above the main diagonal for k positive by k below the main diagonal for k negative. k=0 (the default) gives the main diagonal. Note that the input matrix need not be square."

(maybe the offset language could be worded better. NumPy says "the collection of elements of the form ``a[i, i+offset]``")

Reply:
What about "Returns the kth diagonal of self as a tuple containing all `self[i, j]` for which `j - i = k`, in order of increasing `i + j`, starting with `i + j = |k|`. The main diagonal corresponds to `k=0`; diagonals above and below correspond to `k > 0` and `k < 0`, respectively."?

Reply:
That reads fine to me. 

I was going to comment that it should return a matrix vector. For sparse matrices, it definitely should be sparse, since the shape of a sparse matrix can be potentially very large.